File: main/kernel/src/org/ametys/runtime/group/Group.java
/*
 *  Copyright 2012 Anyware Services
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package org.ametys.runtime.group;

import java.util.HashSet;
import java.util.Set;

/**
 * A user group is a set of logins, representing a group of users.<br>
 * A group contains any number of users, and a single user may belong to any number of groups.
 */
public class Group
{
    private Set<String> _users;
    private String _id;
    private String _label;
    
    /**
     * Constructor.
     * @param id the unique id of this profile
     * @param label the label of this group
     */
    public Group(String id, String label)
    {
        _id = id;
        _label = label;
        _users = new HashSet<String>();
    }
    
    /**
     * Returns the unique Id of this group
     * @return the unique Id of this group
     */
    public String getId()
    {
        return _id;
    }
    
    /**
     * Returns the label of this group
     * @return the label of this group
     */
    public String getLabel()
    {
        return _label;
    }
    
    /**
     * Set the label of this group
     * @param label The new label of the group
     */
    public void setLabel(String label)
    {
        _label = label;
    }
    
    /**
     * Adds an user to this group
     * @param login the login of the user to add to this group
     */
    public void addUser(String login)
    {
        _users.add(login);
    }
    
    /**
     * Removes an user to this group
     * @param login the login of the user to remove to this group
     */
    public void removeUser(String login)
    {
        _users.remove(login);
    }
    
    /**
     * Returns all users of this group.
     * @return Users as a Set of String (login).
     */
    public Set<String> getUsers()
    {
        return _users;
    }
    
    @Override
    public String toString()
    {
        StringBuffer sb = new StringBuffer("UserGroup[");
        sb.append(_id);
        sb.append(" (");
        sb.append(_label);
        sb.append(") => ");
        sb.append(_users.toString());
        sb.append("]");
        return sb.toString();
    }    
    
    @Override
    public boolean equals(Object another)
    {
        if (another == null || !(another instanceof Group))
        {
            return false;
        }
        
        Group otherGroup = (Group) another;
        
        return _id != null  || _id.equals(otherGroup.getId());
    }
    
    @Override
    public int hashCode()
    {
        return _id.hashCode();
    }
}

File: main/plugin-core-ui/resources/js/Ametys/form/widget/Group.js
/*
 *  Copyright 2013 Anyware Services
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

/**
 * This class provides a widget to query and select one or more groups.
 * 
 * This widget is registered for fields of type Ametys.form.WidgetManager#TYPE_STRING.<br>
 */
Ext.define('Ametys.form.widget.Group', {
    extend: 'Ametys.form.AbstractQueryableComboBox',
	
	valueField: 'value',
	displayField: 'displayName',
	
    getStore: function()
    {
        if (!Ext.data.schema.Schema.get('default').hasEntity('Ametys.form.widget.Group.Groups')) {
            Ext.define("Ametys.form.widget.Group.Groups", {
                extend: 'Ext.data.Model',
                
                fields: [
                     {name: 'label', mapping: 'label', type: 'string', sortType: Ext.data.SortTypes.asNonAccentedUCString},
                     {name: 'groupId', mapping: 'id'},
                     {name: 'groupDirectory', mapping: 'groupDirectory'},
                     {name: 'groupDirectoryLabel', mapping: 'groupDirectoryLabel', sortType: Ext.data.SortTypes.asNonAccentedUCString},
                     {
                        name: 'value', 
                        type: 'string',
                        calculate: function(data)
                        {
                            return data.groupId + '#' + data.groupDirectory;
                        }
                     },
                     {
                         name: 'displayName',
                         type: 'string',
                         sortType: Ext.data.SortTypes.asNonAccentedUCString,
                         calculate: function(data)
                         {
                            return data.label + ' (' + data.groupId + ', ' + data.groupDirectoryLabel + ')';
                         }
                     }
                ]
            });
        }

        return Ext.create('Ext.data.Store', {
            model: 'Ametys.form.widget.Group.Groups',
            proxy: {
                type: 'ametys',
                plugin: 'core',
    			url: "groups/search.json", 
                reader: {
                    type: 'json',
                    rootProperty: 'groups'
                }
            },
            
            pageSize: this.maxResult,
            sortOnLoad: true,
            sorters: [{property: 'label', direction:'ASC'}],
            
            listeners: {
                beforeload: {fn: this._onStoreBeforeLoad, scope: this}
            }
        });
    },
    
    /**
     * Set the request parameters before loading the store.
     * @param {Ext.data.Store} store The store.
     * @param {Ext.data.operation.Operation} operation The Ext.data.Operation object that will be passed to the Proxy to load the Store.
     * @private
     */
    _onStoreBeforeLoad: function(store, operation)
    {
        operation.setParams(Ext.apply(operation.getParams() || {}, {
            criteria: operation.getParams().query,
            id: operation.getParams().id ? operation.getParams().id.split(',') : null,
            count: this.maxResult, 
            offset: 0,
            contexts: Ametys.getAppParameter('populationContexts')
        }));
    },
    
    getValue: function ()
    {
    	var value = this.callParent(arguments),
    		convertedValues = [];
    	
    	value = Ext.Array.from(value);
    	
    	Ext.Array.forEach (value, function (val) {
    		var parts = val.split('#');
    		convertedValues.push({
    			groupId: parts[0],
    			groupDirectory: parts[1]
    		});
    	});
    	
    	return this.multiple ? convertedValues : convertedValues[0];
    },
    
    setValue: function (value)
    {
    	value = Ext.Array.from(value);
    	
    	var convertedValues = [];
    	Ext.Array.forEach (value, function (item) {
    		if (Ext.isObject(item))
    		{
    			convertedValues.push(item.id + '#' + item.groupDirectory);
    		}
    		else
    		{
    			convertedValues.push(item);
    		}
    	});
    	
    	this.callParent([convertedValues]);
    },
    
    getSubmitValue: function ()
    {
    	var value = this.getValue();
    	return value != null ? Ext.encode(this.getValue()) : null;
    },
    
    getLabelTpl: function ()
    {
    	var tpl = [];
    	tpl.push('<img width="16" height="16" src="' + Ametys.getPluginResourcesPrefix('core-ui') + '/img/groups/group_16.png"/>');
    	tpl.push('{' + this.displayField + '}');
    	return tpl;
    }
});
